src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/resthandler/RestRetryFailedManagedIndex.kt

```diff
@@ -0,0 +1,168 @@
+    }
+
+    override fun getName(): String {
+        return "retry_failed_managed_index"
+    }
+
+    override fun prepareRequest(request: RestRequest, client: NodeClient): RestChannelConsumer {
+        val indices: Array<String>? = Strings.splitStringByCommaToArray(request.param("index"))
+        if (indices == null || indices.isEmpty()) {
+            throw IllegalArgumentException("Missing indices")
+        }
+        val body = if (request.hasContent()) {
+            XContentHelper.convertToMap(request.requiredContent(), false, request.xContentType).v2()
+        } else {
+            mapOf()
+        }
+
+        val strictExpandIndicesOptions = IndicesOptions.strictExpand()
+
+        val clusterStateRequest = ClusterStateRequest()
+        clusterStateRequest.clear()
+            .indices(*indices)
+            .metaData(true)
+            .local(request.paramAsBoolean("local", clusterStateRequest.local()))
+            .masterNodeTimeout(request.paramAsTime("master_timeout", clusterStateRequest.masterNodeTimeout()))
+            .indicesOptions(strictExpandIndicesOptions)
+
+        return RestChannelConsumer { client.admin().cluster().state(clusterStateRequest, IndexDestinationHandler(client, it, body["state"] as String?)) }
+    }
+
+    private fun isFailedState(managedIndexMetaData: ManagedIndexMetaData): Boolean {
+        return managedIndexMetaData.failed == true
+    }
+
+    inner class IndexDestinationHandler(
+        private val client: NodeClient,
+        channel: RestChannel,
+        private val startState: String?,
+        private val failedIndices: MutableList<FailedIndex> = mutableListOf(),
+        private val listOfIndexMetadata: MutableList<Pair<Index, ManagedIndexMetaData>> = mutableListOf()
+    ) : RestActionListener<ClusterStateResponse>(channel) {
+        override fun processResponse(clusterStateResponse: ClusterStateResponse) {
+            val state = clusterStateResponse.state
+            populateList(state, startState)
+
+            val builder = channel.newBuilder().startObject()
+            if (listOfIndexMetadata.isNotEmpty()) {
+                val updateManagedIndexMetaDataRequest = UpdateManagedIndexMetaDataRequest(listOfIndexMetadata)
+
+                try {
+                    client.execute(UpdateManagedIndexMetaDataAction, updateManagedIndexMetaDataRequest,
+                        object : RestResponseListener<AcknowledgedResponse>(channel) {
+                            override fun buildResponse(acknowledgedResponse: AcknowledgedResponse): RestResponse {
+                                if (acknowledgedResponse.isAcknowledged) {
+                                    builder.field(UPDATED_INDICES, listOfIndexMetadata.size)
+                                } else {
+                                    failedIndices.addAll(listOfIndexMetadata.map {
+                                        FailedIndex(it.first.name, it.first.uuid, "failed to update IndexMetadata")
```

Comment:
Seems like we have a combination of Indexmetadata and index metadata in messages, let's use IndexMetaData in all of them.